### DISCUSSION TITLE
Ideas for improvement

### DISCUSSION THREAD

**Discussion by ssditt** [dawidchyrzynski/arduino-home-assistant#63] [Ideas]
**Great job, Dawid! You helped me a lot!**
Let me share with you some ideas for improvement:
1. Have you applied **Abbreviated topic names** for discovery config? Maybe it could save some bytes of code and memory.
2. Missing **Numbers** as HA device, supported by MQTT discovery. 
3. **Callback function** calls for devices would provide device object address - done in Switch, but missing in Fan - to allow callBack processing in cycle when you have more object of the same device (ideally overloading callBack function assignment to keep compability)
4. Usage of special national characters (with diacritics) in names - using them, the config string is not correct probably due to calculated length of config string/payload?
5. The last topic is related to HA side: I have problem with memory overload on Arduino Uno, where I have 5 switches and I need to add 8 temperature sensors. Just declaration of more than 2 sensors causes program failure. The idea is to use just one sensor and merge all temp values into one payload, separated by e.g. semicolon. Would you know how to split the merged payload across more sensors at HA side? Template or script?

**Comment by ssditt** [comment]
I have already solution for the topic nr. 5. It could help those who struggle with lack of resources on Arduino to use just one sensor on HA, send merged paylod for more of them and split it at HA side.

**Comment by pvoncken** [comment]
Topic nr 3 explains why i have troubles with my multi FAN project.
I'm abusing the FAN's device to control 3 dimmable terace-heaters but could not get the callback to work properly.
Do you know a work around to write 3  analog outputs with this library?

**Reply by ssditt** [reply]
see reply bellow...

**Comment by ssditt** [comment]
Arduino Uno ( I do not share whole code as there is a lot specific features for triac phase control):
1. Standalone HA object decalaration:
  HAFan ob("O", HAFan::SpeedsFeature);
  HAFan ku("K", HAFan::SpeedsFeature);
  HAFan lu("U", HAFan::SpeedsFeature);
  HAFan lz("Z", HAFan::SpeedsFeature);

2. Because callBack does not provide HA object address, there are individual callBacks for each HA object (each for Switch and Speed), forward declarartion:
void s0(bool);
void v0(uint16_t);
void s1(bool);
void v1(uint16_t);
void s2(bool);
void v2(uint16_t);
void s3(bool);
void v3(uint16_t);

3. Structured data for fancoils to save memory
struct _fancoils {
    HAFan *unit;
    const char * name;
    int pin;
    uint16_t speed;
    bool state;
    void (*cbState)(bool);
    void (*cbSpeed)(uint16_t);
};
#define FANCOIL_NUM 4
_fancoils fancoil[FANCOIL_NUM] = {
    { &ob,"Obyvak",5,0,false,s0,v0 },
    { &ku,"Kuchyn",6,0,false,s1,v1 },
    { &lu,"Loznice U",7,0,false,s2,v2 },
    { &lz,"Lozn Z",8,0,false,s3,v3 }
};

4. CallBacks just to get the value (think about to add a bool semafor to indicate value update in loop() ):
void s0(bool state) {
    fancoil[0].state=state;
}
void v0(uint16_t speed) {
        fancoil[0].speed=speed;
}
void s1(bool state) {
    fancoil[1].state=state;
}
void v1(uint16_t speed) {
        fancoil[1].speed=speed;
 }
void s2(bool state) {
    fancoil[2].state=state;
}
void v2(uint16_t speed) {
        fancoil[2].speed=speed;
}
void s3(bool state) {
    fancoil[3].state=state;
}
void v3(uint16_t speed) {
        fancoil[3].speed=speed;
}
5. Fancoils initialization in Setup() (not all features are set here, I prefer to modify them manually in HA later on)
     for(int i=0;i<FANCOIL_NUM;i++) {
        fancoil[i].unit->setName(fancoil[i].name);
        fancoil[i].unit->setRetain(true);
        fancoil[i].unit->setSpeedRangeMax(speedResolution);
        fancoil[i].unit->onStateChanged(fancoil[i].cbState);
        fancoil[i].unit->onSpeedChanged(fancoil[i].cbSpeed);
        pinMode(fancoil[i].pin,OUTPUT);
        digitalWrite(fancoil[i].pin,0);
    }

6. All the others have been processed within loop(). You can add a bool value to each fancoil as a semafor to know when a value has been updated to process the update in loop() ( I use interrupt due to phase triac control).

Allready in production, stable! Major problem I had was a relay switching motoric valve. Snubber across relay contacts is a must.

**Reply by pvoncken** [reply]
I'm going to try that right away.
Thanxs a lot

**Reply by pvoncken** [reply]
Your answer helped me in the right direction.
I made a few changes to the HAFan.h and the HAFan.cpp files from the library.
Now i can use the library and use multiple Fans in the same way i use multiple switches.
Thanxs a lot.
If you're interested i can share the file with you. 

The only strange thing is that i can only use some of th PWM pins for analogWrite on my MEGA-board. PINS 4,8,12,13 work. the others do not.
But i cannot imagine that this is related to the library and therefore of-topic here.

**Reply by ssditt** [reply]
No idea about malfunction of PWM pins@Mega board. As the PWM control is not accurate due to PWM and power phase offset (probably not problem with a heater), I prefer phase synchronised control. Just to get zero crossing signal to a GPIO to generate interrupt and control the output triac based on a counter.  You can have whole-wave power control using zero-crossing optocoupler triac for heater, or using non-zero-crossing optocoupler triac and a timer to split half-wave by timer interrupts. This will allow you to dim lights or motors. Works fine on Arduino Uno. Great result with faster boards - with ESP32 smooth dimming 0-100% by 1% step.

**Reply by ssditt** [reply]
But it will be nice of you to share modification of the HAfan object. I guess you pass a object identificator to callBack?

**Reply by pvoncken** [reply]
I did indeed pass an object identificator to callBack? (HAFan*)
Maybe my coding is not so elegant, but I'm new in Arduino world and for now I'm happy with the result
[HAFan.h.txt](https://github.com/dawidchyrzynski/arduino-home-assistant/files/8017981/HAFan.h.txt)
[HAHVAC.cpp.txt](https://github.com/dawidchyrzynski/arduino-home-assistant/files/8017982/HAHVAC.cpp.txt)
.

**Reply by pvoncken** [reply]
Sorry, I noticed that I attached the wrong file (HAHVAC).
I will fix that tomorrow.

**Reply by pvoncken** [reply]
[HAFan.cpp.txt](https://github.com/dawidchyrzynski/arduino-home-assistant/files/8021734/HAFan.cpp.txt)
[HAFan.h.txt](https://github.com/dawidchyrzynski/arduino-home-assistant/files/8021735/HAFan.h.txt)
